components/scream/src/control/surface_coupling.cpp

```diff
@@ -268,43 +364,57 @@ get_col_info(const std::shared_ptr<const FieldHeader>& fh,
   const auto& dims = layout.dims();
 
   auto lt = get_layout_type(layout.tags());
-  const bool scalar = lt==LayoutType::Scalar2D || lt==LayoutType::Scalar3D;
-  const bool vector = lt==LayoutType::Vector2D || lt==LayoutType::Vector3D;
+  const bool scalar   = lt==LayoutType::Scalar2D || lt==LayoutType::Scalar3D;
+  const bool vector   = lt==LayoutType::Vector2D || lt==LayoutType::Vector3D;
+  const bool layout3d = lt==LayoutType::Scalar3D || lt==LayoutType::Vector3D;
   EKAT_REQUIRE_MSG(scalar || vector,
       "Error! Support for tensor fields not yet implemented in surface coupling.\n");
   EKAT_REQUIRE_MSG( (vecComp<0) == scalar,
       "Error! You can and must specify a vector component only for vector fields.\n");
 
   vecComp = std::max(0,vecComp);
 
-  // Product of dimensions of field, without counting space dims, that is,
-  // the product of dimensions of the "analytical" field.
-  int dims_prod = 1;
-
-  // Compute initial offset based on this field rank and requested component
+  // Compute initial offset. We use get_last_extent() as it
+  // accounts for padding (if any).
   col_offset = vecComp;
-  if (vector) {
-    dims_prod = dims[1];
+  if (layout3d) {
+    if (lt==LayoutType::Vector3D) {
+      col_offset *= fh->get_alloc_properties().get_last_extent();;
+    }
+    col_offset += m_num_levs-1;
   }
 
   // If rank>1, there always is some stride
   col_stride = 1;
   if (layout.rank()>1) {
-    // Note: use alloc prop's last extent, cause it includes padding (if any);
-    col_stride = fh->get_alloc_properties().get_last_extent();
+    col_stride = fh->get_alloc_properties().get_last_extent();;
     if (layout.rank()>2) {
       col_stride *= dims[1];
     }
   }
 
+
   // Recursively go through parent field, adding offset depending
   // on the location of the slice.
   // Note: as of now, I don't expect there to be "a grandparent" for this field,
   //       so we could avoid the while loop. However, the logic is so simple, that
   //       I may as well add it, and be safe down the road if it happens.
+
+  // Product of dimensions of field, without counting space dims, that is,
+  // the product of dimensions of the "analytical" field.
+  int dims_prod = 1;
+    if (vector) {
+      dims_prod = dims[1];
+    }
+
   std::shared_ptr<const FieldHeader> me = fh;
   std::shared_ptr<const FieldHeader> p  = me->get_parent().lock();
   while (p!=nullptr) {
+
+    // TODO: Currently this is not used as we require fields to not have parents in
```

Comment:
Do you think we should just remove this code? I mean, if we don't support subfields, we might as well remove code that we'll not run. It is not that complicated, so we'll probably be able to recreate it if need be...

Reply:
Alternatively it would be simple enough to remove the "no parent" requirement and give the SC-user the option of just handing the child field with no vector component (now that the offset is correct). I think it may be too trivial not for me to try real quick. If it becomes a hassle I'll just delete this section of code.